optimize/scalar: add Brent's method for univariate minimization

imohiyoko opened the pull request:
This PR addresses issue #483 by introducing a new sub-package `optimize/scalar` and implementing **Brent's method** for finding the local minimum of a univariate function.

### Summary of Changes
- Created `gonum/optimize/scalar` package.
- Implemented `BrentMin`, a derivative-free minimization algorithm combining the golden section search and successive parabolic interpolation.
- The implementation is a direct port of **Netlib's `fmin.f`** (by Richard Brent) to ensure numerical stability and correctness.
- Added comprehensive tests in `brent_min_test.go`.

### Design Decisions
- **File Naming**: I chose `brent_min.go` (instead of just `brent.go`) to explicitly distinguish it from Brent's root-finding method (Zeroin). This leaves room to add the root-finding algorithm (discussed in #925) as `brent_root.go` in the same package in the future.
- **API**: The function signature accepts a `Settings` struct for fine-grained control (tolerance, max iterations) and returns a `Result` struct with metadata (status, iteration count), following the convention of the `optimize` package.

### Reference
- Netlib `fmin`: http://www.netlib.org/opt/fmin.f
- Brent, R. P. (1973). *Algorithms for Minimization without Derivatives*.

---
**Note for reviewers:**
I am a university student studying numerical analysis, and this is my first contribution to Gonum (and OSS in general). I have tried to follow the project's conventions, but I would greatly appreciate any feedback or guidance if there are areas for improvement.

Commit:
optimize/scalar: add Brent's method for univariate minimization

kortschak commented on optimize/scalar/brent_min.go:
```diff
@@ -0,0 +1,217 @@
+// Reference: Brent, Richard P. Algorithms for minimization without derivatives. Courier Corporation, 2013.
+func BrentMin(f func(float64) float64, min, max float64, settings *Settings) (Result, error) {
+	// Machine epsilon for float64 (approx 2.22e-16)
+	const epsilon = 1.0 / (1 << 52)
```
```suggestion
	const epsilon = 0x1p-52
```

kortschak commented on optimize/scalar/brent_min.go:
```diff
@@ -0,0 +1,217 @@
+	const c = 0.3819660112501051517954131656343618822796908201942371
+
+	// Default settings
+	tol := math.Sqrt(epsilon)
```
This can be a const.

kortschak commented on optimize/scalar/brent_min_test.go:
```diff
@@ -0,0 +1,107 @@
+		settings   *Settings
+		wantX      float64
+		wantStatus Status
+		distTol    float64 // 許容する x の誤差 (wantX との距離)
```
Comments in English please.

kortschak commented on optimize/scalar/brent_min.go:
```diff
@@ -0,0 +1,217 @@
+	Tol float64
+
+	// MaxIterations is the maximum number of iterations allowed.
+	// If MaxIterations <= 0, a default safety limit (e.g., 100) is used.
```
```suggestion
	// If MaxIterations <= 0, a default safety limit of 100 is used.
```

kortschak reviewed (commented):
Initial review only.

imohiyoko committed:
Update optimize/scalar/brent_min.go

Thanks, applied

Co-authored-by: Dan Kortschak <dan@kortschak.io>

imohiyoko committed:
Update optimize/scalar/brent_min.go

Thanks, applied

Co-authored-by: Dan Kortschak <dan@kortschak.io>

Commit:
optimize/scalar: use hex float literals and const default tolerance

Update epsilon to use the hexadecimal floating-point literal 0x1p-52.
Define the default tolerance (sqrt of epsilon) as a constant 0x1p-26 instead of calculating it at runtime.

Commit:
optimize/scalar: translate comments and use hex literals

- Translate Japanese comments in tests to English.
- Use hexadecimal floating-point literals for epsilon.
- Clarify documentation for default MaxIterations.

kortschak commented on optimize/scalar/brent_min.go:
```diff
@@ -0,0 +1,220 @@
+	Failure
+)
+
+func (s Status) String() string {
```
Instead of this, we can use 
```
//go:generate go tool golang.org/x/tools/cmd/stringer -type Status
```
above the type definition.
```
// Status represents the status of the optimization.
//
//go:generate go tool golang.org/x/tools/cmd/stringer -type Status
type Status int
```

kortschak commented on optimize/scalar/brent_min.go:
```diff
@@ -0,0 +1,220 @@
+	const sqrtEpsilon = 0x1p-26
+
+	// c is the squared inverse of the golden ratio: (3 - sqrt(5))/2
+	const c = 0.3819660112501051517954131656343618822796908201942371
```
```suggestion
	const (
		// Machine epsilon for float64: 2**-52
		epsilon = 0x1p-52
		// Sqrt of machine epsilon: 2**-26
		sqrtEpsilon = 0x1p-26

		// c is the squared inverse of the golden ratio: (3 - sqrt(5))/2
		c = 0.3819660112501051517954131656343618822796908201942371
	)
```

kortschak commented on optimize/scalar/brent_min.go:
```diff
@@ -0,0 +1,220 @@
+	w := x
+	v := x
+	e := 0.0 // step size of the step before last
+	d := 0.0 // step size of the last step
```
```suggestion
	var (
		e float64 // step size of the step before last
		d float64 // step size of the last step
	)
```

kortschak commented on optimize/scalar/brent_min.go:
```diff
@@ -0,0 +1,220 @@
+
+	// Main loop
+	for {
+		iterations++
```
```suggestion
	for i:= 1; ; i++ {
```

kortschak commented on optimize/scalar/brent_min.go:
```diff
@@ -0,0 +1,220 @@
+	for {
+		iterations++
+
+		// Check iteration limit
```
```suggestion
```
The code already says this; see the Comments section of https://doc.cat-v.org/bell_labs/pikestyle.

kortschak commented on optimize/scalar/brent_min.go:
```diff
@@ -0,0 +1,220 @@
+
+		xm := 0.5 * (a + b)
+		tol1 := epsilon*math.Abs(x) + tol/3.0
+		tol2 := 2.0 * tol1
```
```suggestion
		xm := (a + b) / 2
		tol1 := epsilon*math.Abs(x) + tol/3
		tol2 := 2 * tol1
```

kortschak commented on optimize/scalar/brent_min.go:
```diff
@@ -0,0 +1,220 @@
+		tol2 := 2.0 * tol1
+
+		// Check stopping criterion
+		if math.Abs(x-xm) <= (tol2 - 0.5*(b-a)) {
```
```suggestion
		if math.Abs(x-xm) <= (tol2 - (b-a)/2) {
```

kortschak commented on optimize/scalar/brent_min.go:
```diff
@@ -0,0 +1,220 @@
+			r := (x - w) * (fx - fv)
+			q := (x - v) * (fx - fw)
+			p := (x-v)*q - (x-w)*r
+			q = 2.0 * (q - r)
```
```suggestion
			q = 2 * (q - r)
```

kortschak commented on optimize/scalar/brent_min.go:
```diff
@@ -0,0 +1,220 @@
+			e = d
+
+			// Check acceptability of the parabolic fit
+			if math.Abs(p) >= math.Abs(0.5*q*etemp) || p <= q*(a-x) || p >= q*(b-x) {
```
```suggestion
			if math.Abs(p) >= math.Abs(q*etemp/2) || p <= q*(a-x) || p >= q*(b-x) {
```

kortschak commented on optimize/scalar/brent_min_test.go:
```diff
@@ -0,0 +1,107 @@
+				t.Fatalf("unexpected error: %v", err)
+			}
+
+			// Check status
```
```suggestion
```

kortschak commented on optimize/scalar/brent_min_test.go:
```diff
@@ -0,0 +1,107 @@
+				t.Errorf("status mismatch: got %v, want %v", res.Status, tc.wantStatus)
+			}
+
+			// Check value precision for converged cases
```
```suggestion
```

kortschak commented on optimize/scalar/brent_min_test.go:
```diff
@@ -0,0 +1,107 @@
+			min:        0,
+			max:        5,
+			settings:   nil, // Default settings
+			wantX:      2.0,
```
```suggestion
			wantX:      2,
```

kortschak commented on optimize/scalar/brent_min_test.go:
```diff
@@ -0,0 +1,107 @@
+			min:        0,
+			max:        5,
+			settings:   &Settings{Tol: 1e-10}, // Stricter tolerance
+			wantX:      2.0,
```
```suggestion
			wantX:      2,
```

kortschak commented on optimize/scalar/brent_min_test.go:
```diff
@@ -0,0 +1,107 @@
+			max: 5,
+			// Set a very low iteration limit to ensure it stops early
+			settings:   &Settings{MaxIterations: 2},
+			wantX:      2.0,            // Ignored in check
```
```suggestion
			wantX:      2,            // Ignored in check
```

kortschak commented on optimize/scalar/brent_min_test.go:
```diff
@@ -0,0 +1,107 @@
+			settings:   &Settings{MaxIterations: 2},
+			wantX:      2.0,            // Ignored in check
+			wantStatus: IterationLimit, // Should hit limit
+			distTol:    10.0,           // Ignored in check
```
```suggestion
			distTol:    10,           // Ignored in check
```

kortschak commented on optimize/scalar/brent_min_test.go:
```diff
@@ -0,0 +1,107 @@
+		distTol    float64 // Tolerance for the distance from wantX
+	}{
+		{
+			name: "Parabola: (x-2)^2 + 1",
```
```suggestion
			name: "parabola",
```
The names don't need to fully describe the test, just be uniquely identifying. Similarly below.

kortschak commented on optimize/scalar/brent_min_test.go:
```diff
@@ -0,0 +1,107 @@
+			distTol:    1e-7,
+		},
+		{
+			name: "Strict Tolerance: (x-2)^2 + 1",
```
```suggestion
			name: "strict_tolerance",
```
Names have spaces replaced with underscore, so by doing that in advance, it becomes easier to copy the failing name and find the test in the source.

kortschak commented on optimize/scalar/brent_min_test.go:
```diff
@@ -0,0 +1,107 @@
+			distTol:    1e-9, // Expect higher precision
+		},
+		{
+			name: "Iteration Limit Check",
```
```suggestion
			name: "iteration_limit_check",
```

kortschak commented on optimize/scalar/brent_min_test.go:
```diff
@@ -0,0 +1,107 @@
+)
+
+func TestBrentMin(t *testing.T) {
+	tests := []struct {
```
Hoist this to a package var `brentMinTests`.

kortschak commented on optimize/scalar/brent_min_test.go:
```diff
@@ -0,0 +1,107 @@
+		},
+	}
+
+	for _, tc := range tests {
```
```suggestion
	for _, test := range brentMinTests {
```
with required changes below.

Commit:
"optimize/scalar: use stringer for Status"

imohiyoko committed:
Update optimize/scalar/brent_min.go

Co-authored-by: Dan Kortschak <dan@kortschak.io>

imohiyoko committed:
Update optimize/scalar/brent_min.go

Co-authored-by: Dan Kortschak <dan@kortschak.io>

imohiyoko committed:
Update optimize/scalar/brent_min.go

Co-authored-by: Dan Kortschak <dan@kortschak.io>

imohiyoko committed:
Update optimize/scalar/brent_min.go

Co-authored-by: Dan Kortschak <dan@kortschak.io>

imohiyoko committed:
Update optimize/scalar/brent_min.go

Co-authored-by: Dan Kortschak <dan@kortschak.io>

imohiyoko committed:
Update optimize/scalar/brent_min.go

Co-authored-by: Dan Kortschak <dan@kortschak.io>

imohiyoko committed:
Update optimize/scalar/brent_min_test.go

Co-authored-by: Dan Kortschak <dan@kortschak.io>

imohiyoko committed:
Update optimize/scalar/brent_min.go

Co-authored-by: Dan Kortschak <dan@kortschak.io>

imohiyoko committed:
Update optimize/scalar/brent_min.go

Co-authored-by: Dan Kortschak <dan@kortschak.io>

imohiyoko committed:
Update optimize/scalar/brent_min_test.go

Co-authored-by: Dan Kortschak <dan@kortschak.io>

imohiyoko committed:
Update optimize/scalar/brent_min_test.go

Co-authored-by: Dan Kortschak <dan@kortschak.io>

imohiyoko committed:
Update optimize/scalar/brent_min_test.go

Co-authored-by: Dan Kortschak <dan@kortschak.io>

imohiyoko committed:
Update optimize/scalar/brent_min_test.go

Co-authored-by: Dan Kortschak <dan@kortschak.io>

imohiyoko committed:
Update optimize/scalar/brent_min_test.go

Co-authored-by: Dan Kortschak <dan@kortschak.io>

Commit:
optimize/scalar: refactor BrentMin to address review feedback

- Group constants and state variables (e, d) into blocks.
- Simplify numeric expressions and literals (e.g., use /2 instead of *0.5).
- Use hexadecimal floating-point literals for epsilon.
- Prepare Status type for stringer and remove manual String method.

imohiyoko committed:
Update optimize/scalar/brent_min_test.go

Co-authored-by: Dan Kortschak <dan@kortschak.io>

imohiyoko committed:
Update optimize/scalar/brent_min_test.go

Co-authored-by: Dan Kortschak <dan@kortschak.io>

imohiyoko committed:
Update optimize/scalar/brent_min_test.go

Co-authored-by: Dan Kortschak <dan@kortschak.io>

imohiyoko committed:
Update optimize/scalar/brent_min_test.go

Co-authored-by: Dan Kortschak <dan@kortschak.io>

Commit:
optimize/scalar: implement BrentMin with refactored logic and tests

Updated the implementation to align with Gonum coding standards:
- Improved constant/variable grouping and numeric expressions.
- Integrated stringer for Status type.
- Refactored tests by hoisting cases and simplifying identifiers.

kortschak commented on optimize/scalar/brent_min_test.go:
```diff
@@ -0,0 +1,106 @@
+	"testing"
+)
+
+// brentMinTests is hoisted to a package variable as suggested by the reviewer.
```
```suggestion
```

kortschak commented on optimize/scalar/brent_min.go:
```diff
@@ -0,0 +1,206 @@
+	fv := fx
+
+	for i := 1; ; i++ {
+
```
```suggestion
```

imohiyoko committed:
Update optimize/scalar/brent_min.go

Co-authored-by: Dan Kortschak <dan@kortschak.io>

imohiyoko committed:
Update optimize/scalar/brent_min_test.go

Co-authored-by: Dan Kortschak <dan@kortschak.io>
